r/src/compute-exec.cpp

```diff
@@ -125,6 +125,46 @@ std::shared_ptr<arrow::Schema> ExecNode_output_schema(
   return node->output_schema();
 }
 
+// [[arrow::export]]
+std::string ExecPlan_ToString(const std::shared_ptr<compute::ExecPlan>& plan) {
+  return plan->ToString();
+}
+
+// [[arrow::export]]
+std::string ExecPlan_ToStringWithSink(
+    const std::shared_ptr<compute::ExecPlan>& plan,
+    const std::shared_ptr<compute::ExecNode>& final_node, cpp11::list sort_options,
+    int64_t head = -1) {
```

Comment:
I created `ExecPlan_prepare()`, but I haven't updated `ExecPlan_run()`. Do you think this would be a good split (not sure how much of your PR I should incorporate in mine)?

Reply:
Yep, so pretty much what I had before as `ExecPlan_ToStringWithSink` (without the actually ToString) becomes a separate function.

Reply:
> (But make sure you can actually use it to eliminate the duplication in `ExecPlan_run()`!)

That's the part I am struggling with at the moment.

Reply:
That's what it looks like. Also, depending on which part we separate into `ExecPlan_prepare`, we might need to return more than 1 object (which, by default, would be the plan), for it to work with `ExecPlan_run`.

Reply:
Yep, that's one of the things I wanted to try today / this morning. Not quite sure what is the type for the sink node (in compute-exec.cpp I believe `sync_gen` is the sync node) - is that correct?

Reply:
Would it be `std::shared_ptr<compute::ExecBatch>`?

Reply:
Nope, I think it needs to be an `ExecNode` since that's what `MakeExecNodeOrStop` returns.

Reply:
I think this looks very similar to my approach in https://github.com/apache/arrow/pull/13541/commits/f40e5d289553543ce070ac67bbcf565b289548b3.

Reply:
I don't have formed thoughts on the cost of the tradeoff, but definitely the variant with `ExecPlan_prepare` looks cleaner and avoids the _double accounting_. 

On the [CI failure](https://github.com/apache/arrow/runs/7445803518?check_suite_focus=true#step:7:26180): I don't think it has anything to do with this PR since it seems present in most PR and was probably introduced in the last couple of days.

Reply:
I did. I forgot to submit my comment.